File: src/main/java/io/github/alex_hawks/util/Vector3.java
package io.github.alex_hawks.util;

public class Vector3
{
    public final int x, y, z;
    
    public Vector3(int x, int y, int z)
    {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector3(double x, double y, double z)
    {
        this.x = (int) Math.round(x);
        this.y = (int) Math.round(y);
        this.z = (int) Math.round(z);
    }

    @Override
    public int hashCode()
    {
        final int prime = 31;
        int result = 1;
        result = prime * result + x;
        result = prime * result + y;
        result = prime * result + z;
        return result;
    }

    @Override
    public boolean equals(Object obj)
    {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Vector3 other = (Vector3) obj;
        if (x != other.x)
            return false;
        if (y != other.y)
            return false;
        if (z != other.z)
            return false;
        return true;
    }

    @Override
    public String toString()
    {
        return "Vector3 [x=" + x + ", y=" + y + ", z=" + z + "]";
    }
}

File: src/main/scala/io/github/alex_hawks/util/minecraft/common/Vector3.scala
package io.github.alex_hawks.util.minecraft.common

import net.minecraft.util.EnumFacing
import net.minecraft.util.math.{AxisAlignedBB, BlockPos}

class Vector3(val x: Int, val y: Int, val z: Int) {

  def this(pos: BlockPos) = this(pos.getX, pos.getY, pos.getZ)

  def +(w: (Int, Int, Int)): Vector3 = new Vector3(x + w._1, y + w._2, z + w._3)

  def +(that: Vector3): Vector3 = new Vector3(this.x + that.x, this.y + that.y, this.z + that.z)

  override def toString = s"V3(${x},${y},${z})"

  def canEqual(other: Any): Boolean = other.isInstanceOf[Vector3]

  override def equals(other: Any): Boolean = other match {
    case that: Vector3 =>
      (that canEqual this) &&
        x == that.x &&
        y == that.y &&
        z == that.z
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(x, y, z)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }

  def toPos: BlockPos = new BlockPos(this.x, this.y, this.z)

  def shift(dir: EnumFacing): Vector3 = this +(dir.getFrontOffsetX, dir.getFrontOffsetY, dir.getFrontOffsetZ)

  def shiftAABB(aabb: AxisAlignedBB):AxisAlignedBB = {
    import aabb._
    new AxisAlignedBB(minX + x, minY + y, minZ + z, maxX + x, maxY + y, maxZ + z)
  }


}
